app/models/spree/address.rb

```diff
@@ -10,12 +10,16 @@ class Address < ActiveRecord::Base
     has_one :enterprise, dependent: :restrict_with_exception
     has_many :shipments
 
+    before_validation :geocode, if: :use_geocoder?
+
     validates :firstname, :lastname, :address1, :city, :country, presence: true
     validates :zipcode, presence: true, if: :require_zipcode?
     validates :phone, presence: true, if: :require_phone?
 
     validate :state_validate
 
+    attr_accessor :use_geocoder
```

Comment:
I'm actually wondering if we could remove the automatic geocoding from the model code completely. If all places where you change the address have geocoding in the browser with a map to confirm then we don't need this. Less complexity, especially around AR callbacks, is very valuable.

Reply:
@mkllnk Yes we don't have this new map with location confirm feature in the Admin > Enterprise > Address section yet. So until we do I had to add this new `use_geocoder` attribute and leave the geocoding code in the `Spree::Address` model.

This is because every time someone changes their address in the Admin > Enterprise > Address section the address would automatically be geocoded so this could overwrite the latitude/longitude that was set during sign up. To prevent the latitude/longitude from being overwritten this add's a checkbox which people need to explicity click if they want their address to be automatically geocoded, otherwise it will just use the manually configured latitude/longitude.

![new-use-geocoder-setting](https://user-images.githubusercontent.com/101088/107782827-6e8d7c00-6d41-11eb-8af6-e1a88f5241b3.png)

Reply:
Yeah I really don't like this logic moving to the Address model. Adding these form fields and the checkbox is great :point_up: 

I'll take a look and try to give some more useful feedback...

Reply:
Alright, there's two places where we want to do some geocoding on enterprise addresses, and that's on enterprise creation in `app/controllers/api/enterprises_controller.rb` around line 16, and when an enterprise is edited, which is `app/controllers/admin/enterprises_controller.rb` around line 50.

I think I would remove these all bits like `attr_accessor :use_geocoder` from the Address class and add a service which can be called from those two places above. It can take the params that have been posted, and those can include the value of that checkbox. It can hold the various bits of conditional logic we obviously want here, like not geocoding if that checkbox is unchecked, or if the longitude and latitude have been passed manually on registration. We can call it after the enterprise is successfully saved in both cases.

Reply:
@Matt-Yorkley sounds good to me. I have extracted out a `AddressGeocoder` service now and added support for the `use_geocoder` parameter to the `Api::EnterprisesController`, please see 955d0d3